staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -132,21 +137,89 @@ func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.
 	}
 	err = s.openAPIVersionedService.RegisterOpenAPIVersionedService("/openapi/v2", pathHandler)
 	if err != nil {
-		return nil, err
+		klog.Fatalf("Failed to register versioned open api spec for root: %v", err)
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.OpenAPIV3) {
+		s.openAPIV3VersionedService, err = handler3.NewOpenAPIService(nil)
+		if err != nil {
+			return nil, err
+		}
+		err = s.openAPIV3VersionedService.RegisterOpenAPIV3VersionedService("/openapi/v3", pathHandler)
+		if err != nil {
+			return nil, err
+		}
+
+		i = 1
+		for delegate := delegationTarget; delegate != nil; delegate = delegate.NextDelegate() {
+			handler := delegate.UnprotectedHandler()
+			if handler == nil {
+				continue
+			}
+
+			apiServiceName := fmt.Sprintf(localDelegateChainNamePattern, i)
+			s.UpdateOpenAPIV3(*downloader, apiServiceName, handler)
+			i++
+		}
 	}
 
 	return s, nil
 }
 
+func (s *specAggregator) UpdateOpenAPIV3(downloader Downloader, apiServiceName string, handler http.Handler) error {
+	s.openAPIV3rwMutex.Lock()
+	defer s.openAPIV3rwMutex.Unlock()
+	groups, err := downloader.DownloadV3(handler)
+	if err != nil {
+		return err
+	}
+	if _, exists := s.openAPIV3Specs[apiServiceName]; !exists {
+		s.openAPIV3Specs[apiServiceName] = make(map[string]*openAPIV3SpecInfo)
+	}
+
+	// Remove any groups that do not exist anymore
+	for group := range s.openAPIV3Specs[apiServiceName] {
+		if _, exists := groups[group]; !exists {
+			s.openAPIV3VersionedService.DeleteGroupVersion(group)
+		}
+		delete(s.openAPIV3Specs[apiServiceName], group)
+	}
+
+	for group, info := range groups {
+		if info.spec == nil {
+			continue
+		}
+
+		// If ETag has not changed, no update is necessary
+		oldInfo, exists := s.openAPIV3Specs[apiServiceName][group]
+		if exists && oldInfo.etag == info.etag {
+			continue
+		}
+		s.openAPIV3Specs[apiServiceName][group] = &openAPIV3SpecInfo{
+			spec: info.spec,
+			etag: info.etag,
+		}
+		s.openAPIV3VersionedService.UpdateGroupVersion(group, info.spec)
+	}
+	return nil
+}
+
 type specAggregator struct {
-	// mutex protects all members of this struct.
+	// mutex protects all openAPI v2 members of this struct.
 	rwMutex sync.RWMutex
 
 	// Map of API Services' OpenAPI specs by their name
 	openAPISpecs map[string]*openAPISpecInfo
 
 	// provided for dynamic OpenAPI spec
 	openAPIVersionedService *handler.OpenAPIService
+
+	// mutex protects all openAPI v3 members of this struct.
+	openAPIV3rwMutex sync.RWMutex
+
+	openAPIV3Specs map[string]map[string]*openAPIV3SpecInfo
```

Comment:
spec by what?

Reply:
By usage I guess `ByGroupVersionAndGroup`, very odd. Why twice by group?

Reply:
APIService and then by GroupVersion. Updated the documentation